File: src/types/Condition.ts
import * as DB from "tsondb/schema/dsl"
import { NestedTranslationMap } from "./Locale.js"
import { Errata } from "./source/_Erratum.js"
import { src } from "./source/_PublicationRef.js"

export const Condition = DB.Entity(import.meta.url, {
  name: "Condition",
  namePlural: "Conditions",
  type: () =>
    DB.Object({
      src,
      translations: NestedTranslationMap(
        DB.Required,
        "Condition",
        DB.Object({
          name: DB.Required({
            comment: "The condition’s name.",
            type: DB.String({ minLength: 1 }),
          }),
          rules: DB.Optional({
            comment: "Additional rules for the condition, if applicable.",
            type: DB.String({ minLength: 1, isMarkdown: true }),
          }),
          effects: DB.Required({
            comment: "The effects for level 1 to 4.",
            type: DB.Array(DB.String({ minLength: 1, isMarkdown: true }), {
              minItems: 4,
              maxItems: 4,
            }),
          }),
          errata: DB.Optional({
            type: DB.IncludeIdentifier(Errata),
          }),
        }),
      ),
    }),
  instanceDisplayName: {},
  uniqueConstraints: [
    {
      entityMapKeyPath: "translations",
      keyPathInEntityMap: "name",
    },
  ],
})

File: Sources/OptolithDatabaseSchema/Types/Condition.swift
import FileDB

@Model
public struct Condition {
    /// The publications where you can find the entry.
    @MinItems(1)
    let src: [PublicationRef]

    /// All translations for the entry, identified by IETF language tag (BCP47).
    @Relationship(Locale.self)
    let translations: [String: Translation]

    @Embedded
    struct Translation {  // ConditionTranslation
        /// The condition’s name.
        @MinLength(1)
        let name: String

        /// Additional rules for the condition, if applicable.
        @MinLength(1)
        @Markdown
        let rules: String?

        /// The effects for level 1 to 4.
        @MinItems(4)
        @MaxItems(4)
        let effects: [ConditionLevelEffect]

        /// A list of errata for the entry in the specific language.
        @MinItems(1)
        let errata: [Erratum]?
    }
}

@TypeAlias
public struct ConditionLevelEffect {
    @MinLength(1)
    @Markdown
    let text: String
}
